### DISCUSSION TITLE
Recommended API mocking setup

### DISCUSSION THREAD

**Discussion by SplittyDev** [vercel/next.js#49371] [Help]
### Summary

I'm using NextJS 13.4 (App Router) for my frontend, while the backend is being developed separately (long-running NodeJS server). The backend isn't ready yet, so I am mocking the API using `msw`.

Using `msw` for mocking does (somewhat) work, but setting it up properly for SSR is a bit of a nightmare, since it needs separate setups for SSR and CSR, and there's no "global" entry point where I could set up mocking.

I just realized that there's an even bigger problem with `msw`: It doesn't work at all in middleware, since as far as I can tell, that's being rendered using the `edge` runtime.

There are other mocking solutions such as axios mock adapter, but NextJS docs recommend using `fetch` for caching purposes.

Is there a good way to mock API requests in SSR, CSR and middleware contexts?

**Comment by maorazoulay** [comment]
Were you able to figure it out? I tried using mirageJs but I am not sure where to instantiate the server since there is no entry point such as index.js (I am using the App router)

**Reply by condinoaljoseph** [reply]
Hi how its going? Im using json-server with json-server-auth right now. How about you guys?

**Reply by ganicus** [reply]
There's another ongoing discussion with the MSW folks and Next team posted in this repo's discussions section. The issue has to do with how integral patching global fetch is the the app router framework and that that 3rd party patching fetch doesn't persist reliably.

How have your other solutions worked out so far? Do you have code examples?

#53409

**Reply by SalahAdDin** [reply]
Were you able to make it work? @SplittyDev
